stdlib/public/runtime/DynamicCast.cpp

```diff
@@ -705,6 +728,56 @@ tryCastToSet(
   return DynamicCastResult::Failure;
 }
 
+static DynamicCastResult
+tryCastToString(
+  OpaqueValue *destLocation, const Metadata *destType,
+  OpaqueValue *srcValue, const Metadata *srcType,
+  const Metadata *&destFailureType, const Metadata *&srcFailureType,
+  bool takeOnSuccess, bool mayDeferChecks)
+{
+  assert(srcType != destType);
+  assert(destType->getKind() == MetadataKind::Struct);
+  const auto destStructType = cast<StructMetadata>(destType);
+  assert(destStructType->Description == &NOMINAL_TYPE_DESCR_SYM(SS));
+
+  switch (srcType->getKind()) {
+  case MetadataKind::ObjCClassWrapper: { // Obj-C -> String
+#if SWIFT_OBJC_INTEROP
+    // Call _tryCastFromClassToObjCBridgeable with memoized witness lookups.
+    static const _ObjectiveCBridgeableWitnessTable *destBridgeWitness;
+    static const Metadata *targetBridgedClass;
+    static swift_once_t fetchWitnessOnce;
+    swift_once(&fetchWitnessOnce,
+               [](void *_destType) {
+                 auto destType = (const Metadata *)_destType;
+                 destBridgeWitness = findBridgeWitness(destType);
+                 targetBridgedClass = _getBridgedObjectiveCType(
+                   MetadataState::Complete, destType, destBridgeWitness).Value;
+               }, (void *)destType);
+    if (srcType == targetBridgedClass) {
```

Comment:
Right now, this logic only kicks in if the `srcType` is exactly the same class that `String` expects to bridge from.  I suspect this can be generalized a bit.

Reply:
@Catfish-Man and @milseman might have better ideas?

Reply:
I think I worked out the logic for checking whether the `srcType` is an ObjC subclass of the expected bridge type.  Testing shows this gets us onto the new early-bridging path for all `NSString` objects, including `NSMutableString` and `CFString`.